### PR TITLE
test set integration

### PR DESCRIPTION
**Pull request by Victorlouisdg**

### CODE REVIEW

File: keypoint_detection/data/datamodule.py [hunk 1]
```diff
@@ -51,4 +71,11 @@ def val_dataloader(self):
         return dataloader
 
     def test_dataloader(self):
-        pass
+        dataloader = DataLoader(
+            self.test_dataset,
+            min(4, self.batch_size),
```
Comment (tlpss) [inline keypoint_detection/data/datamodule.py:76]: magic number -> document

File: keypoint_detection/models/detector.py [hunk 2]
```diff
@@ -180,7 +185,7 @@ def configure_optimizers(self):
             },
         }
 
-    def shared_step(self, batch, batch_idx, is_validation_step=False) -> Dict[str, Any]:
+    def shared_step(self, batch, batch_idx, include_visualization_data=False) -> Dict[str, Any]:
```
Comment (tlpss) [inline keypoint_detection/models/detector.py:188]: `include_visualization_data_in_result_dict`  + add to docstring

File: keypoint_detection/utils/visualization.py [hunk 3]
```diff
@@ -54,38 +53,52 @@ def visualize_predictions(
...
-            keypoint_channel = f"{keypoint_channel[0]}+{keypoint_channel[1]}+..."
-    keypoint_channel_short = (keypoint_channel[:40] + "...") if len(keypoint_channel) > 40 else keypoint_channel
-    label = f"{keypoint_channel_short}_{mode}_keypoints"
-    logger.experiment.log(
-        {label: wandb.Image(grid, caption="top: predicted heatmaps, middle: predicted keypoints, bottom: gt heatmap")}
-    )
+    return grid
+
+
+if __name__ == "__main__":
```
[the remark this answers is not in the document]
Comment (tlpss) [reply keypoint_detection/utils/visualization.py:76]: + add functionality to readme?

File: keypoint_detection/utils/visualization.py [hunk 4]
```diff
@@ -54,38 +53,52 @@ def visualize_predictions(
     imgs: torch.Tensor,
     predicted_heatmaps: torch.Tensor,
     gt_heatmaps: torch.Tensor,
-    logger: pytorch_lightning.loggers.WandbLogger,
```
Comment (tlpss) [inline keypoint_detection/utils/visualization.py:57]: docstring?

File: keypoint_detection/train/train.py [hunk 5]
```diff
@@ -88,7 +103,27 @@ def parse_channel_configuration(channel_configuration: str) -> List[List[str]]:
...
+            If not specified, the train dataset will be split to create a validation set.",
+    )
+    datasets_group.add_argument(
+        "--json_test_dataset_path",
+        type=str,
+        help="Absolute path to the json file that defines the test dataset according to the COCO format. \
+            If not specified, no test set evaluation will be performed at the end of training.",
+    )
+
     parser = KeypointsDataModule.add_argparse_args(parser)
```
Comment (tlpss) [inline keypoint_detection/train/train.py:127]: Maybe have datamodule take the 3 strings and instantiate them? then the argparse arguments can be moved to the DataModule as well as the logic for dealing with the optional datasets

File: keypoint_detection/train/train.py [hunk 6]
```diff
@@ -54,8 +54,19 @@ def main(hparams: dict) -> Tuple[KeypointDetector, pl.Trainer]:
...
+
+    if "json_validation_dataset_path" in hparams:
+        validation_dataset = COCOKeypointsDataset(
+            hparams["json_validation_dataset_path"], hparams["keypoint_channel_configuration"]
+        )
+    if "json_test_dataset_path" in hparams:
+        test_dataset = COCOKeypointsDataset(
+            hparams["json_test_dataset_path"], hparams["keypoint_channel_configuration"]
+        )
+
```
Comment (tlpss) [inline keypoint_detection/train/train.py:68]: move to Datamodule

**Review by tlpss — CHANGES_REQUESTED** [review]

**Review by tlpss — APPROVED** [review]
